### ISSUE TITLE
`secretRef` is not working for Grafana Loki

### ISSUE THREAD

**Issue by Pavaningithub** [fjogeleit/policy-reporter#314]
While sending logs to Grafana Loki, using `secretRef` giving below error.

ArgoCD Application manifest:
```yaml
project: default
source:
  repoURL: 'https://kyverno.github.io/policy-reporter'
  targetRevision: 2.18.3
  helm:
    values: |-
      ui:
        enabled: true
      metrics:
        enabled: true
      rest:
        enabled: true
      target:
        loki:
          secretRef: "kyverno-policy-reporter-target-loki"
          minimumPriority: "warning"
  chart: policy-reporter
destination:
  server: 'https://kubernetes.default.svc/'
  namespace: kyverno-policy-reporter
syncPolicy:
  automated:
    prune: true
    selfHeal: true
  syncOptions:
    - CreateNamespace=true
```

**Error:**
Viewing kyverno-policy-reporter pod log
```bash
$ > kubectl logs kyverno-policy-reporter-5855f49cb-rmbrm -f
1.6850868947274184e+09  warn    failed to get secret reference  {"error": "Get \"https://x.x.x.x:443/api/v1/namespaces/kyverno-policy-reporter/secrets/kyverno-policy-reporter-target-loki\": dial tcp x.x.x.x:443: connect: connection refused"}
1.6850868947274933e+09  info    UI configured
1.685086894735476e+09   info    REST api enabled
1.6850868947355518e+09  info    metrics enabled
1.6850868947356982e+09  info    start client    {"worker": 5}
```

1. `kyverno-policy-reporter-target-loki` exists in the `kyverno-policy-reporter` namespace.
2. No NetworkPolicy defined.
3. Kyverno policy reporter RBAC is not tampered.

**Comment by fjogeleit** [comment]
thanks for reporting, I will have a look on it.

**Comment by Pavaningithub** [comment]
> thanks for reporting, I will have a look on it.

This is running on EKS cluster.

**Comment by fjogeleit** [comment]
Thanks sorry I had not much time last week, I try to test it today

**Comment by fjogeleit** [comment]
Hm so I tested it with available k3s cluster and it worked, unfortunately i don't have an EKS cluster to test with. Did you restart the pod? Does it always fail?

One option you could try as an alternative are `mountedSecrets`. Its not documented yet - sorry for that. Its basically the some functionality but instead you mount the secret as volume. This was introduced as contribution especially to support the usage of AWS KMS.

Your secret need the fields as JSON. You provide the path of the secret in the values as follows:

https://github.com/kyverno/policy-reporter/blob/main/charts/policy-reporter/values.yaml#L297

Volumes can be added in the values.yaml as follows:

https://github.com/kyverno/policy-reporter/blob/main/charts/policy-reporter/values.yaml#L655

The related PR / Context:

https://github.com/kyverno/policy-reporter/pull/282

**Comment by Pavaningithub** [comment]
Referring to PR #282: `mountedSecrets` is only present under `s3` target and missed in all other targets in `charts/policy-reporter/config.yaml`

**Comment by justinwalz** [comment]
It looks like it fails on the first request (my guess is istio-proxy is not ready yet). The code could be more robust with a retry, or the end-user could delay the startup of the kyverno-policy-reporter container to wait. 


Log line

```
2023-05-30 17:36:50     WARN    failed to get secret reference  {"error": "Get \"https://172.20.0.1:443/api/v1/namespaces/kyverno-policy-reporter/secrets/kyverno-policy-reporter-target-loki\": dial tcp 172.20.0.1:443: connect: connection refused"}
```

Probably failing here:
- https://github.com/kyverno/policy-reporter/blob/main/pkg/config/target_factory.go#L869
- https://github.com/kyverno/policy-reporter/blob/main/pkg/kubernetes/secrets/client.go#L37

**Comment by fjogeleit** [comment]
> Referring to PR #282: `mountedSecrets` is only present under `s3` target and missed in all other targets in `charts/policy-reporter/config.yaml`

Hm yes you are right, the mapping of the vlaues is missing for other outputs. I will fix that.

**Comment by fjogeleit** [comment]
@justinwalz thanks for coming up with this, you also right, that could be a reason. I will work on a retry mechanism with a maybe configurable delay between them.

**Comment by fjogeleit** [comment]
I tagged a v2.15.2 version with a retry mechanism. Would be helpful if you could test it by setting image.Tag to 2.15.2 if it works I will release a new Helm chart version where I also fixed the mountedSecret mapping in the config

**Comment by justinwalz** [comment]
Thanks for the quick fix! 

I tried updating the image, but `ghcr.io/kyverno/policy-reporter:2.15.2` doesn't exist (yet).

**Comment by fjogeleit** [comment]
Yes, build takes some time but should be available now

**Comment by justinwalz** [comment]
Got it - tested, and it works. Thank you.

**Comment by fjogeleit** [comment]
Thanks. I will try to release it today

**Comment by fjogeleit** [comment]
@justinwalz @Pavaningithub I released v2.19.2 of the helm chart with the missing `mountedSecret` mapping as well as the retry mechanism for `secretRef`.

Let me know if your issues are solved and I can close the issue or feel free to do it yourself.

**Comment by fjogeleit** [comment]
I close the issue, let me know if still something does not work as expected.
